### ISSUE TITLE
[Discussion/Planning] Implement bidirectional scrolling in ecash

### ISSUE THREAD

**Issue by jboniface** [Expensify/Expensify.cash#2985]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

# Problem
We're trying to enable comment linking in Cash, and while developing the solution, we've hit a snag : in short, we're unsure how to load the specific report comment. At the moment, because of pagination when the `fetchReportActions` we’re just fetching 50 results. So if the report had 300 comments and we want the comment link to go to comment sequence number 100 it gets tricky because we have two options. 

1. Have fetch report actions load actions 50 to 100 than 250 to 300. Or,
2. Load 100 to 300 actions.

If we go with option 2 it would be extremely slow since we load bottom to top so it would be a bad user experience to load the desired comment action last.

Option 1 sounds like the better user experience but then our current FlatList only loads more comments if you scroll up. This means the user can scroll up to load report comments between 1-50 but they will never be able to scroll down to load report comments between 100 - 150.

# Solution

Create a PR to build bidirectional scrolling so that we can load comments onStart or onEnd and then we continue with comment linking. This way even if the comment is pretty old the user will be able to scroll up and down to load additional comments using pagination.

# Discussion
As per above, we've [decided](https://expensify.slack.com/archives/C01GTK53T8Q/p1620176991391300) we want to build our own FlatList component to support bidirectional scrolling, but we need to plot out the actual approach for implementation.

cc @chiragsalian

**Comment by MelvinBot** [comment]
Triggered auto assignment to @michaelhaxhiu (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by roryabraham** [comment]
cc @marcaaron I know you showed me a gist or draft PR a while back with an outline of an implementation. Do you still have that?

**Comment by kidroca** [comment]
Do you know the total size of chat items before or when the chat loads?
E.g. do you know you have a million messages in a chat or you it's unknown

>  but they will never be able to scroll down to load report comments between 100 - 150.

What's the problem with starting in the middle and then append/prepend to the FlatList - you can insert more items to the bottom when you start to reach the bottom or insert more items to the top when you start to reach the top

**Comment by roryabraham** [comment]
> Do you know the total size of chat items before or when the chat loads?

I'm 95% sure we could easily provide this information from the API.

**Comment by marcaaron** [comment]
Yes, we already know since the entire `reportActionList` gets downloaded when the app inits 🙃 
(but the pagination and storage of messages happen in `Report_GetHistory`)

**Comment by marcaaron** [comment]
>What's the problem with starting in the middle and then append/prepend to the FlatList

append works fine since there is an `onEndReached` method and we already use it for the one direction

prepend not so much for a couple of reasons:

1. There is no `onStartReached` - make could create our own by wrapping `FlatList` and looking at the scroll metrics to detect when the `ScrollView` has reached the "top" (or bottom rather in the inverted style)
2. When prepending items the content shifts. We might be able to solve this if we knew the total height of all items that we are prepending and could offset the scroll position. Perhaps there is a simpler way or something I'm missing.

**Comment by marcaaron** [comment]
This is something new I just noticed in the past few months that looks interesting

https://github.com/GetStream/react-native-bidirectional-infinite-scroll

Have no idea if it supports web - but I think we are basically trying to achieve the same thing as this package.

**Comment by kidroca** [comment]
This library seems to support everything you need at the moment 
- onStartReached
- onEndReached
- virtualization
- maintain correct scroll position
- potentially fix the a11y and content being copied in reverse if you don't use the inverted version

It also supports the same virtualization props as the FlatList
> This package is a wrapper around react-native's FlatList. So it accepts all the props from FlatList

Is it possible to create a new ChatList component based on that library and use it for Android and iOS, while desktop and web continue to use the existing FlatList (ReportActionsView) implementation.
When web support is added - the old FlatList can be dropped altogether

The point is if you try to create your own solution you'll basically create the same thing and have the same problems e.g. make it work for web

**Comment by marcaaron** [comment]
I can't speak for everyone, but I'm not sure if we would commit to using a library on the mobile side with only the _hope_ that the web side would work out for us eventually (maybe if we can also test the web PR in progress and deem that it's heading in the right direction). 

In any case, the first step would be to evaluate if such a library exists and can actually work for us (cross platform or not).

>The point is if you try to create your own solution you'll basically create the same thing and have the same problems e.g. make it work for web

An alternative way to frame this is that if we create our own solution, we might have the same problems, but nothing can stop us from solving them.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @stitesExpensify (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @tgolen (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by kidroca** [comment]
What are the limits for this ticket. Can we use a library?

@marcaaron The library that you found seems to be the real deal
If it doesn't work for web, we could tailor just the web implementation ourselves - it's steal beats trying to cover all platforms ourselves. And in fact that's exactly what's the PR you've shared does - for web they are working on a standalone web solution that is not using a FlatList but exposes the same interface, and is virtualized, but using [react-virtuoso](https://github.com/petyosi/react-virtuoso) for virtualization

> Create a PR to build bidirectional scrolling...

I think this is a leftover from when the ticket was internal, as an external contributor, I can spend a lot of time to do a PR only to be rejected. 

How should this ticket be handled?
It's hard making proposal without starting to work on a bi-direction list, but it's also very time consuming to do so

**Comment by kidroca** [comment]
Using the library and making a web specific implementation (that we can contribute to that library) can be a lot less coding: 
see their attempt: https://github.com/GetStream/react-native-bidirectional-infinite-scroll/blob/a782467396b3143b43ff2ca15b4ada3f46a7da20/src/BidirectionalFlatList.tsx ~300 lines of code 

This is because we have much more flexibility coding for web

**Comment by roryabraham** [comment]
@kidroca I think what you've suggested sounds pretty reasonable. I would love to see a solution that forks https://github.com/GetStream/react-native-bidirectional-infinite-scroll and implements a shim to make it compatible with react-native-web. We can use that fork ourselves and try to get it merged into the parent repo. Alternatively, we can just create the shim directly in Expensify.cash but I would prefer the first option.

Given your history of excellent work, I would feel comfortable hiring you for that task. Then if you produce a proof-of-concept and we decide that we'd still rather implement a bidirectional paginated FlatList from scratch, then you'd still be paid for that work. It would still be valuable research to push this issue forward.

cc @jboniface @marcaaron @chiragsalian – sound reasonable to you?

**Comment by quinthar** [comment]
Yes, I love the idea of forking and contributing a RN4W version of https://github.com/GetStream/react-native-bidirectional-infinite-scroll.  Great!

**Comment by marcaaron** [comment]
As for proposing a web implementation it might be worth inquiring on the status of this PR -> https://github.com/GetStream/react-native-bidirectional-infinite-scroll/pull/16

Maybe we can build on that work, propose an alternative, or help get it prioritized.

But verifying if this project suits our needs sounds like a good first step.

**Comment by kidroca** [comment]
Sounds great!
They are indeed working on another PR for web, so that's good news as well.

@roryabraham 
I guess you'd want to see how this lib will integrate with E.cash so you can decide is it worth it to continue with it. Did I get that right?

I can make a POC implementation where we can see how it goes for **native** and make further plans if it's worth it 

Setting:
- a chat with 300+ messages 
- load the last 50 messages
- jump to message 100th -> load 25-30 messages above and below
- then scrolling up or down should load and  append data correctly for both directions and not cause any flashing or jerk

**Comment by roryabraham** [comment]
@kidroca Sounds great as a first step. So let's do that first and then discuss.

However, I think it's also in-scope for this job to test the web beta (i.e: point E.cash at the latest commit on [this PR](https://github.com/GetStream/react-native-bidirectional-infinite-scroll/pull/16))? If it doesn't work and we can create a reproducible test case, then that's valuable information to provide to the PR author.

Let me know if you have any other questions.

**Comment by kidroca** [comment]
> However, I think it's also in-scope for this job to test the web beta (i.e: point E.cash at the latest commit on [this PR](https://github.com/GetStream/react-native-bidirectional-infinite-scroll/pull/16))? If it doesn't work and we can create a reproducible test case, then that's valuable information to provide to the PR author.

Good idea
Can we just make the scope of the current task clear.
I'll will use the latest code (from the PR) to make a POC, it should work for web, but _"If it doesn't work and we can create a reproducible test case"_ - I'm not sure what that involves and who's expected to do it, but it sounds like extra work 
- is it a part of this ticket?
- create a reproducible test case -> does this mean to crate a react-native project that demonstrates the problem? (Expo snack might do the trick as well: https://snack.expo.io/)

**Comment by jboniface** [comment]
@kidroca tbh, I was going to raise the value on this job right before you accepted it, and it does sound like more work than i originally proposed, so I'll just throw a bonus on and double the payout for you when it's done. Just link back to this comment if I forget somehow and we'll figure it out.

**Comment by jboniface** [comment]
So to answer your question:

- is it a part of this ticket? 

Yes, but I'll pay you more for it

**Comment by roryabraham** [comment]
> If it doesn't work and we can create a reproducible test case

Ideally this would be part of the ticket, but I would say it's not a hard requirement. It's tough to comment on this because I don't know how complex the potential issue would be to diagnose and/or reproduce. If there is some very complex or major issue, then maybe that would be a sign we don't want to try and move forward with the library. But if it's something basic with a simple reproduction, then ideally we'll work with the contributor of that library towards a solution, and creating a minimal reproduction in an Expo snack would probably be a valuable contribution for them to test a fix against.

Does that answer your question? I'm hoping for this can be a collaborative process between you, us at Expensify, and the author of that 3rd-party library.

**Comment by kidroca** [comment]
Thanks, I think you can assign me the ticket in gitlab as well

**Comment by kidroca** [comment]
@jboniface @roryabraham 
I've Submitted a PR with the POC implemented using https://github.com/GetStream/react-native-bidirectional-infinite-scroll and the PR that supports web

Short summary
- it renders on `web` but has a lot of issues
- it has a bug on `react-native` where sometimes while you hit the top of the list and it should load past content - it does but you are moved to the bottom of the list
  - it might be possible this is the same issue: https://github.com/GetStream/react-native-bidirectional-infinite-scroll/issues/8
- starting from a past message could be made to work better, it might seem a bit slow ATM as everything is hidden by a fullscreen loader until content is ready

More details in the PR #3705

**Comment by roryabraham** [comment]
Very tangentially related, I noticed that [this PR](https://github.com/facebook/react-native/pull/29466) (opened almost a year ago), got approved today. I wonder if that might mean that https://github.com/GetStream/flat-list-mvcp can soon be deprecated.

**Comment by kidroca** [comment]
Good find!
This prop is indeed the foundation for the bidirectional list

Together with the new better documented [viewabilityconfig](https://reactnative.dev/docs/flatlist#viewabilityconfig)
it could allow enough flexibility to achieve what we want in e.cash land 

Working on this ticket I've thought of 2 alternatives based on either VirtualList or FlatList 

#### 1. Rely solely on virtualization
Basic idea: Since we know how many total messages a chat has - `maxSequenceNumber` we can leverage the `VirtualizedList` [getitemcount](https://reactnative.dev/docs/virtualizedlist#required-getitemcount) and [getItem](https://reactnative.dev/docs/virtualizedlist#required-getitem)
- `getItemCount` returns `maxSequenceNumber` 
- `getItem` returns a `reportAction` 
- when `getItem` cannot return an action we render a loading state
  - if we have to - we start a process that would resolve the item

This pattern relies on `getItemLayout` being implemented.

#### 2. Use boundary items that work as waypoints 
This strategy provides an `onTopReached` and `onBottomReached` in a regular FlatList 

- We have a list of items
- Before passing it to the FlatList 
- We add one *special* item to the beginning and another one to the *end* 
- as we scroll the list and a *special* item is getting near viewable range we append another batch of items to the list
  - we can also use the `componentDidMount` of a *special* item as it will only get mounted when we scroll close to it 

This pattern is called Waypointing and is used in infinite lists

**Comment by kidroca** [comment]
> Very tangentially related, I noticed that [this PR](https://github.com/facebook/react-native/pull/29466) (opened almost a year ago), got approved today. I wonder if that might mean that https://github.com/GetStream/flat-list-mvcp can soon be deprecated.

This would make `maintainVisibleContentPosition` available in `iOS` and `Android` but would need to be implemented for `web` in `react-native-web` as well

**Comment by roryabraham** [comment]
Okay, well I'm not sure which is the better of those two alternatives, so let's try to list the pros/cons of each. I'll give my best effort, but feel free to give your interpretation of this analysis too:

### Rely solely on virtualization
#### Pros
- Potentially a simpler implementation. I think that this approach (not using waypoints) is probably more approachable to a new engineer looking at the codebase for the first time.

#### Cons
- We won't begin fetching new report actions until we've failed to render one that the user tried to look at (i.e: we try to render a `ReportActionItem` at a position in our list where we know there should be one, but we don't have the data, so we fetch that data). I'm thinking this would not provide the best UX, especially on low-bandwidth connections.

### Use boundary items that work as waypoints
#### Pros
- Easy to control the render boundary by adding waypoints in the middle of a group of loaded chats:
    - Let's say we have chats 50-100 loaded
    - Place waypoints in between chats 25 and 26, and between chats 75 and 76.
    - That way, we still have a margin of 25 chats preloaded before we begin fetching more. And we could easily adjust that threshold.

#### Cons
- Small additional storage requirement for the special marker elements (this might be negligible)
- Might lead to more complex logic around sequence numbers and adding new chats.

---

So by my cursory analysis, it seems like waypoints are a better solution. But I could be missing things, and I'm curious for the inputs of others here @kidroca @tgolen @marcaaron.

**Comment by marcaaron** [comment]
>probably more approachable to a new engineer looking at the codebase for the first time
>Might lead to more complex logic around sequence numbers and adding new chats.

Sorry, I'm only sort of following the discussion here. But I agree that these are good reasons to avoid the waypoint based solution.

**Comment by kidroca** [comment]
### Rely solely on virtualization
#### Pros 
  - you don't have to worry about appending prepending items to the chat
  - you can scroll to any item and it will appear, or load then appear
#### Cons
This have one major requirement: `getItemLayout` 
Scrolling to an `item/index` outside the pre-rendered window requires `getItemLayout` to be implemented as noted here:
https://reactnative.dev/docs/flatlist#scrolltoindex

This is a bit more complicated as we might not even have the item locally in the first place
- when we don't have an item we just return a sane estimate value e.g. `32` 
- then when we have the item we cache it's rendered size the way it's performed for web 
- it's possible to optimize this by storing the calculated layout size in Onyx (in the `actionItem`)

It's not certain what would happen with scroll position at this point - it might get messed up when the items around our anchor render as well. The prop `maintainVisibleContentPosition` would probably help to overcome this 

Another potential problem is some action items should not be rendered, this might mean we simply render `null` for them. But it still gets in the way of layout computation and it becomes more complex

Overall this needs to be tested in a POC to see if it'll work

### Use boundary items that work as waypoints
#### Pros
- simpler to implement as it doesn't need `getItemLayout`

#### Cons
- still needs `maintainVisibleContentPosition` or a way to preserve the scrolling position as more items are added to the top or bottom of the list
  - I'm not very certain on this, as you know from the current flat list appending items to the top of the list (as you scroll up) does not affect scroll position 

---

Some other notes:

> We won't begin fetching new report actions until we've failed to render one

It's might seem that way initially but it's not true
We have a virtual list with a size of 15 screens - one we currently see, 7 pre-rendered above, and below 
When you start to scroll in a direction more items are added to the pre-rendered area that you don't yet see
this is the point that will trigger fetching their content and it might be long ready when you actually scroll to it

This works for waypoints to - you don't actually need to put the waypoint items at 25 and 75, you can put them around the edges of locally available data, as soon as you are within 6-7 pages of that edge - you will know to load more data

> Small additional storage requirement for the special marker elements

There are only 2 markers at the edges, they are (pre) rendered when you get near the edge and they don't render anything complex but a spinner

> Might lead to more complex logic around sequence numbers and adding new chats.

The filter and rendering logic remains the same, we just add 1 special item to the beginning and one special item to then end before passing the array down to the FlatList

> I think that this approach (not using waypoints) is probably more approachable to a new engineer looking at the codebase for the first time.

The waypoint logic can be encapsulated to a component the way you do with the platform implementations for the `InvertedFlatList` 

The new component passes data wrapped by 2 boundary items to the FlatList
It also forwards all the props to the FlatList but exposes 2 more props - `onNearingListTop`, `onNearingListBottom`. 

This way the more complex work of the list is abstracted and you don't immediately need to know it to be able to understand how the app works

--- 

Since this is getting in different direction I would like to complete my current milestone - I've done the work that I was hired for 
If you have further need for me, let's open a new milestone
I won't be engaging in further discussion regarding the alternatives

**Comment by jboniface** [comment]
just noting that rory and i discussed and we agree that you've satisfied the requirements of the milestone. I'm going to pay out and close the job, but we're leaving the issue for further discussion while we determine the best path forward. When it's time to move forward on a solution, I'll file a new issue and flag you @kidroca

**Comment by jboniface** [comment]
Oh, sorry, we're on a payment moratorium while they upgrade our account, I'll need to do this once UW finishes their maintenance. Setting a reminder.

**Comment by kidroca** [comment]
@jboniface 
Thanks,
I see I have a "new" offer regarding this ticket in Upwork, though it's not coming from you but from Omri Benami
Also you did mention a price bump for the ticket: https://github.com/Expensify/Expensify.cash/issues/2985#issuecomment-864071630 since the job is re-oferred can we just apply it on the offer?

**Comment by jboniface** [comment]
@mallenexpensify is Omri our agent at UW? 

Regarding the price bump, I was planning to just tack it on via the bonus field

**Comment by jboniface** [comment]
@kidroca were you able to accept the contract? I think we might be waiting on an action from UW, but I can't action the contract yet.

**Comment by kidroca** [comment]
@jboniface I wasn't sure, so I haven't. I'll go and accept it now

**Comment by jboniface** [comment]
Ok, cool, once you've done that, I think someone on the UW team needs to do something to share the contract with me, so it could be a little while longer before I can pay.

**Comment by kidroca** [comment]
An interesting article on Discord react-native optimization
https://blog.discord.com/how-discord-achieves-native-ios-performance-with-react-native-390c84dcd502#4fc0

Key points: 
The article claims that this library worked fast and almost covered their need back then
https://github.com/Flipkart/recyclerlistview -> covers web as well
Might be worth checking it out since it's still maintained and updated

In the end they've implemented their own native component 
Even though the article makes it sound like they've implemented in entirely on the JS side it's not the case
Also `FastList` it's not published as an open source RN package yet

**Comment by jboniface** [comment]
@roryabraham do you have any thoughts on how we can move this forward in the short term? Threads functionality is currently blocked on this (as well as comment linking in general).

(as a quick aside, the RND work was paid out)

**Comment by MelvinBot** [comment]
@tgolen, @kidroca it looks like no one is assigned to work on this job.
Please double the price or add a comment stating why the job isn't being doubled.

**Comment by roryabraham** [comment]
Despite the fact that we exported some R&D for this issue, I think we'll probably keep this internal going forward.

**Comment by puneetlath** [comment]
Given @roryabraham's last comment, are @tgolen and @kidroca still the right people to be assigned to this issue? Or do we need to find a volunteer to take it on?

**Comment by roryabraham** [comment]
Yep, I'm working on this.

**Comment by jboniface** [comment]
@roryabraham can you update the OP to match the path forward that you've settled on and start providing weekly updates to this issue?

**Comment by roryabraham** [comment]
I am going to work on this for a while today, and I'll post a comprehensive update with my progress.

**Comment by jboniface** [comment]
@roryabraham can we get an update here, please?

**Comment by roryabraham** [comment]
These are the only tangible updates I have so far:

- The react-native-bidirectional-infinite-scroll library is causing our unit tests to fail, and I haven't been able to figure out why, but I did reach out to the library maintainer. Other people are having the same issue.
- Another challenge is that I can't really test bidirectional scrolling without building comment linking or finding some weird hack to test it. So I've been working on figuring that out.
- With the exception of the failing unit tests the bidirectional-scrolling FlatList is working as a drop-in replacement for the unidirectional FlatList we already have

I'm currently working to see how I can advance the issue. For now I'm focusing on getting it working and leaving performance considerations in the back of my mind.

**Comment by kidroca** [comment]
> Another challenge is that I can't really test bidirectional scrolling without building comment linking or finding some weird hack to test it. So I've been working on figuring that out.

Hey @roryabraham 
I think the easiest thing would be to borrow some code from this PR: https://github.com/Expensify/App/pull/3705

https://github.com/Expensify/App/blob/110d3e47d7a1de6bbf8631691a698cc93764922f/src/pages/home/ReportScreen.js#L78-L81

Given that you'll open a conversation of 200 messages 
You can hardcode something like `anchorSequenceNumber={100}` which would start you from the 100th message, providing you with the chance to scroll up or down to load more.

I even hooked this up with dynamic linking so that you can open a chat on a specific message (anhorSequenceNumber in route params)

**Comment by roryabraham** [comment]
Just a quick update – I have been making some good progress in my [draft PR](https://github.com/Expensify/App/pull/4894/files). I have been seeing several of the same issues that @kidroca encountered in his POC, so I am breaking off in a fork of react-native-bidirectional-infinite-scroll and am going to attempt to see if I can fix some issues at the native level.

**Comment by roryabraham** [comment]
Okay, I've spent a lot of time on this today and I'm actually now thinking that https://github.com/Flipkart/recyclerlistview is the most promising option for a few reasons:

- Already has functional web support
- Purports much better performance than standard RN FlatList (this could be a big win in-and-of-itself)
- Seems far more actively maintained and widely-used than https://github.com/GetStream/react-native-bidirectional-infinite-scroll
- I believe it will be possible to implement bidirectional infinite scrolling in `recyclerlistview` without having to write or maintain any new native code. Just TypeScript. 
- Also, it's slightly easier to work on locally because `react-native-bidirectional-infinite-scroll` internally uses a separate package called `react-native-mvcp` where most of the native code lives. My investigation showed that we would probably need to fork both packages to overcome some of the issues I ran into there, and it just seems very challenging / a big PITA.

Here's a couple draft PR's that show the progress I've made on implementing a POC of bidirectional scrolling in `recyclerlistview`:

- https://github.com/Flipkart/recyclerlistview/pull/647
- https://github.com/roryabraham/recyclerlistview-example/pull/1

But there are a few issues I encountered:

1. For some reason, `onScroll` is not being executed (on web only) in my [example app](https://github.com/roryabraham/recyclerlistview-example/pull/1).
    - I created an [issue](https://github.com/Flipkart/recyclerlistview/issues/645) for this, so hopefully someone might be able to help me find out what's going wrong there.
    - I'm not treating this as a dealbreaker because there's a functional example of `recyclerlistview`'s `onEndReached` working with `react-native-web` [here](https://codesandbox.io/s/k54j2zx977). Seems like I'm just holding something wrong or the repo doesn't play nice with Expo, which shouldn't be a deal-breaker for us.
    - If you can see the difference between the functional example and my broken example, let me know! I'm pretty stumped on that!
1. I was also having issues getting the remote JS debugger to work with the example app this evening, which really limited my ability to debug the changes I made in `recyclerlistview`. Hopefully should be able to work this out soon.
1. In my current implementation of `recyclerlistview`, it seems like `onStartReached` is always being immediately executed, instead of only when the user scrolls. Need to dig into why that's happening and what to do about it.

Overall these issues seem less daunting than the ones we encountered with `react-native-bidirectional-infinite-scroll`. The biggest downside I currently see to `recyclerlistview` is that, unlike `react-native-bidirectional-infinite-scroll`, it has a substantially different API than `FlatList`. This will mean that the Expensify/App PR to implement it would be _much_ more substantial, and that all our contributors would have to gain some familiarity with this library (it comes with a bit of a learning curve). That will also mean that if the core react-native library eventually rolls out substantial improvements to their list components, it would be more work to sunset `recyclerlistview` and go back to using vanilla RN.

**Comment by kidroca** [comment]
IMO the `ReportActionsView` would undertake some hefty changes, no matter what, in order to achieve the desired behavior
I agree it would be easier if we don't have to tinker native iOS and Android code
And I see some value in `RecyclerListView` as well

Some notes:

> Purports much better performance than standard RN FlatList

I'm not sure whether we'll be able to receive the performance benefit, as the [#why](https://github.com/Flipkart/recyclerlistview#why) section explains: _"RecyclerListView encourages you to have deterministic heights for items you need to render, ...,  all you need is a way to look at the data and compute height upfront so that RecyclerListView can compute layout in one pass rather than waiting for the draw to happen"_

`FlatList` has the same optimisation, the problem is we can't pre-compute messages heights in advance, we have the `html` content and we only know the size after it's rendered.
We can come up with a close estimate `height` - this should work for both FlatList and the Recycler view

> I believe it will be possible to implement bidirectional infinite scrolling in `recyclerlistview` without having to write or maintain any new native code. Just TypeScript.

I think this brings us back to a place where we have a unidirectional list and we need to extend it with bidirectional functionality.
Can we explore other alternatives (perhaps in parallel) like my "Waypoint" idea which should be applicable on any list 

Another idea that I had is to use a `VirtualizedList` - we can implement [getItemCount](https://reactnative.dev/docs/virtualizedlist#required-getitemcount) - we know this as it's fetched with the report data `reportActionCount` from `reportSummaryList`. We'll be kind-a having all the items in the list but they're only getting loaded as you scroll (or jump) to an item. Perhaps the `RecyclerListView` has an interface for the same thing

**Comment by kidroca** [comment]
What I don't understand about `RecyclerListView` is how they justify the claim that it has better performance and memory efficiency 

E.g. when you scroll a chat to the past in a `FlatList` items on the bottom are getting removed and items on the top are added - you have a fixed memory window. The argument against that is frequent garbage collection and object creation. But I don't get how the `RecyclerListView` handles this better - it too has to release some messages as you scroll to the past as otherwise memory usage will grow indefinitely

**Comment by roryabraham** [comment]
> Another idea that I had is to use a VirtualizedList - we can implement getItemCount - we know this as it's fetched with the report data reportActionCount from reportSummaryList. We'll be kind-a having all the items in the list but they're only getting loaded as you scroll (or jump) to an item. Perhaps the RecyclerListView has an interface for the same thing

I'm not sure I understand this proposed solution – how would you know to begin fetching newer chats as you scroll down?

**Comment by kidroca** [comment]
### The Virtualized List idea

1. We have a `VirtualizedList` that works with abstract data -> `getItemCount` returns the total action items there are for this conversation
2. Let's say we have a chat with 200 messages
3. When the chat loads we can tell the `VirtualizedList` to start on message # 100 (or any message for that matter) `initialScrollIndex={100}` (we'll usually start on the latest message)
4. `getItems` is called with `index: 100` 
5. If we have data about this item in Onyx we use it to render the item (like we currently do). If we don't have it we render a `<Placholder />` or a `<Skeleton />` message - you've seen them: 
![Skeleton data placeholder](https://user-images.githubusercontent.com/12156624/133478808-720f9442-6f9d-46a8-90a0-cf1d61d4c685.png)
And start a request to fetch data about this item, actually a page of items (like we currently do). The data is merged in Onyx which triggers the `<Skeleton />` item to be replaced with the actual ActionItem

As the user scrolls up and down (or even jumps to the begging of the conversation). `getItems` is triggered for items that are outside the immediate view window, just by scrolling we're triggering an item to be loaded 
Since we're fetching data in pages of 50 items, we can start a fetch only when `getItem` renders a placeholder for indexes divisible by `50`. E.g. initially when you open a chat 100 or more items are rendered at once - we don't want to start 100 requests

We can further polish this but that's the gist of it

One caveat that we might have with this approach though is content shifting to the wrong place. This is because scroll position needs to be preserved while the message that we render would push each other up down - this is something that the Bidirectional List seems to be struggling too, and it's one of the reasons to be created actually

<details>
<summary>...</summary>

> > Another idea that I had is to use a `VirtualizedList` - we can implement [getItemCount](https://reactnative.dev/docs/virtualizedlist#required-getitemcount) - we know this as it's fetched with the report data `reportActionCount` from `reportSummaryList`. We'll be kind-a having all the items in the list but they're only getting loaded as you scroll (or jump) to an item. Perhaps the `RecyclerListView` has an interface for the same thing
> 
> I'm not sure I understand this proposed solution – how would you know to begin fetching newer chats as you scroll down?

</details>

**Comment by roryabraham** [comment]
Thanks for explaining @kidroca! That seems like a good idea but I think you're correct that content shifting (aka "scroll jumping" would likely be an issue).

**Comment by kidroca** [comment]
Yep, but I think "scroll jumping" is a problem we need to solve anyway
Unidirectional lists have this problem when you append more items in the "rear" end of the list (depends on inverted) 
Removing or adding items to a list would also cause this issue
Even the bi-directional list has this problem when the conditions are right 😭

**Comment by roryabraham** [comment]
Getting pretty close to a working POC implementation of a bidirectional scrolling list via recyclerlistview. Was hoping to get it working by EOD today but there are still a couple (hopefully minor) bugs I need to address with the current implementation.

- forked [recyclerlistview](https://github.com/Flipkart/recyclerlistview/pull/647)
- [Example RN app](https://github.com/roryabraham/RecyclerListViewExample)

I'm going to try and finish up that POC and then will be presenting the various options to slack to discuss the best plan of action.

**Comment by roryabraham** [comment]
POC is done, check out the work done in the forked repo [here](https://github.com/Flipkart/recyclerlistview/pull/647) and the example app [here](https://github.com/roryabraham/RecyclerListViewExample) 🎉

I'm going to bring my findings to Slack, but if we get consensus that we want to move forward with this solution, then the next steps would be:

1. Fork https://github.com/Flipkart/recyclerlistview in the Expensify GH Org.
1. Create a PR to switch `ReportActionsView` from `FlatList` -> `RecyclerListView`.
    - I could use some help performing a performance analysis of this PR. Maybe @kidroca could assist?
1. Implement comment linking / bidirectional scrolling in this repo.
1. Create several PRs to update the upstream `recyclerlistview` from our fork.
    - One to fix the window resize issue
    - Another to fix the Android `onEndReached` bug (usually not called on Android unless a non-zero `onEndReachedThreshold` is passed).
    - A third to implement bidirectional scrolling.
    - Another to optimize `DataProvider.cloneWithRows` to take a `lastModifiedIndex` in addition to `firstModifiedIndex`
    - Refactor the library to use Reanimated 2 (this one could be tough but would almost certainly help boost performance, not yet implemented)

**Comment by roryabraham** [comment]
> If we have data about this item in Onyx we use it to render the item (like we currently do). If we don't have it we render a <Placeholder /> or a <Skeleton /> message - you've seen them:
Skeleton data placeholder
And start a request to fetch data about this item, actually a page of items (like we currently do). The data is merged in Onyx which triggers the <Skeleton /> item to be replaced with the actual ActionItem

I like this idea. Right now, my implementation will just display a header/footer + block scrolling, but we could pretty easily implement support for a `Skeleton` component like you've suggested.

**Comment by kidroca** [comment]
The example app does not used the forked version of the Recycler - no `onStartReached` 
I'm not sure what will happen when we start to use dynamic size items, ATM it seems to be set to `50`. I know there's an override prop that would let the Recycler handle it, but wouldn't that mess with the `onStartReached` logic?
Excuse me if I'm being a PITA here 🐈‍⬛

**Comment by roryabraham** [comment]
> The example app does not used the forked version of the Recycler - no onStartReached

Oh yeah sorry, forgot to mention this... check out the `release-local` npm script in `package.json` – that's how I was developing locally. Bit of a pain but it works.

> I'm not sure what will happen when we start to use dynamic size items, ATM it seems to be set to 50. I know there's an override prop that would let the Recycler handle it, but wouldn't that mess with the onStartReached logic?

So `RecyclerListView` _does_ work with dynamic size items, albeit with a performance cost, and I think we should be okay. But I haven't actually tested. It's a good idea to edit my example app to provide random-height list items and see if anything breaks apart.

**Comment by roryabraham** [comment]
Provided an update on this [here](https://expensify.slack.com/archives/C01KVUJKQ84/p1633128550105700). Tl;DR I'm going to keep this as a weekly and get PRs up to implement bidirectional scrolling in `recyclerlistview` and in RN's `FlatList`, and see what bites we get between now and whenever this needs to be re-prioritized.

**Comment by roryabraham** [comment]
Hoping to make progress on this today, assuming I can finish out my last daily first.

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @roryabraham eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by trjExpensify** [comment]
👋 Hey @roryabraham! Can we get a quick update on the latest and greatest with this for N7, please?

**Comment by roryabraham** [comment]
@luacmartins Has joined the fray and we made a fair amount of progress today. You can follow along with the status of the implementation [here](https://github.com/Expensify/App/pull/7071)

**Comment by roryabraham** [comment]
Made a lot of progress in [this messy draft PR](https://github.com/facebook/react-native/pull/33094) towards implementing bidirectional scrolling directly in React Native's FlatList. I honestly think that the remaining bugs are related to the implementation of the `maintainVisibleContentPosition` prop in the native layer more than any JS code.

**Comment by roryabraham** [comment]
I went ahead and cleaned up what I've got so far: https://github.com/Expensify/react-native/pull/7. Bumping this to weekly. I've also been talking with @azimgd and would like to create an Upwork job so that he can spend some time trying to tackle the TODOs listed in the PR.

**Comment by mallenexpensify** [comment]
I assigned myself as the CM since I've been following along.  @roryabraham , for the Upwork job, is there a specific deliverable I should mention? A specific amount, or hourly?

**Comment by roryabraham** [comment]
Posted https://github.com/Expensify/App/issues/7925, which _seems_ like it's going to be a blocker for any solution using React Native's `ScrollView`.

**Comment by roryabraham** [comment]
Thanks for your continued work and interest @azimgd. Here's are my thoughts:

Working on this issue has been a humbling experience for me – I've tried many different approaches over several months and unfortunately have not been able to successfully implement the feature. During that time, I have successfully produced several POC's that, like the one you're showing us, work in a sandboxed environment using mock data and simple square components. Each time, I have discovered some unforeseen bug or bugs that surfaced at a later step that invalidated my POC to some degree.

It's totally possible that the solution you've been working on solves all the problems we've faced in implementing bidirectional scrolling in this repo and doesn't introduce other unseen ones, but there's no way I can reliably validate that without implementing your solution in our product. I for one still have greater confidence that fixing [this bug](https://github.com/Expensify/App/issues/7925) and completing the `onStartReached` support directly in React Native's `FlatList` is the best path forward we have.

If you believe that your react-virtual approach can solve this issue on all platforms without introducing other bugs, that is great. But if you want to pursue that route, you'll need to provide not just a proof-of-concept, but a full-fledged implementation of your approach in our app. However, I unfortunately have not seen enough to make me confident that your POC will be more successful than my several others (though I would love to be proven wrong). So given that, I am not currently comfortable assigning this issue to you or agreeing on a price for that work at this time. I will say that, for a full-fledged solution we would be happy to pay at least the sticker price for [this issue](https://github.com/Expensify/App/issues/7925), _provided that we have not hired someone else for that issue at the time the solution is delivered_.

So if you want to keep working on that, I encourage you to do so, as long as you understand that such work is not officially sanctioned at this time, so there is some risk that it won't be used or you won't be paid for it. Hopefully that all makes sense and transparently expresses our position here.

**Comment by mallenexpensify** [comment]
Guess we'll keep this one open for now, the main issue we're addressing now is https://github.com/Expensify/App/issues/7925

**Comment by mallenexpensify** [comment]
Happy anniversary!
<img width="389" alt="image" src="https://user-images.githubusercontent.com/22508304/169094537-0ecd3a62-d8b1-44d6-9b17-ad19fb13c23d.png">

#7925 is moving along (but also on hold...)

**Comment by mallenexpensify** [comment]
Rory "Made some great progress on Android implementation last week, posted update [here](https://expensify.slack.com/archives/C035J5C9FAP/p1655407541640779)"
https://github.com/Expensify/App/issues/7925#issuecomment-1162353996

**Comment by roryabraham** [comment]
Meeting with Margelo today, we are making slow-and-steady progress. Almost all the pieces are in place now

**Comment by mallenexpensify** [comment]
@roryabraham , can you provide a quick update when ya can? (Mostly cuz i'm curious how close we are)

**Comment by roryabraham** [comment]
Bidirectional scrolling features are live in NewDot production. Now we are waiting for @chiragsalian to re-implement comment linking using the new bidirectional scrolling features so we can battle-test them. Last I heard, he was prioritizing finishing API refactors first.

**Comment by chiragsalian** [comment]
Yup its what rory said. I went ahead and closed my previous PR. Mostly because there was a lot that wasn't ideal about it and it wasn't in any super rush like it was when we originally created the issue so it would be nice to do it better this round.

I'm currently focusing on API refactors but will work on this right after. This time i'll focus on doing very small parts at a time so that reviewing and testing becomes a lot easier. 

Additionally i wanted to ask. Do we feel like a design doc is needed to discuss technical implementation using bidirectional scrolling to get more eyes on this or for each task should i just post a plan of action to have it approved before implementing or just implement each task and discuss improvements during review? Any preferences?

**Comment by mallenexpensify** [comment]
Missed your comment @chiragsalian , thanks for the ping @michaelhaxhiu 

Chirag, any update since your last one? 
Looks like the GH for comment linking got bumped back to `Weekly` https://github.com/Expensify/Expensify/issues/147480

>Do we feel like a design doc is needed to discuss technical implementation using bidirectional scrolling to get more eyes 

Has much changed since the comment linking portion in the comment actions design doc was created? It's been a bit. Maybe a mini doc is in order? Maybe a post in #expensify-open-source with CMEs tagged if you/we are unsure? 
https://docs.google.com/document/d/1vfTkJEOqCu0bVKP9ATKhMqAfLslYMuoL6W80lLCB5U4/edit#heading=h.my7t93ltzxp1

**Comment by chiragsalian** [comment]
> Has much changed since the comment linking portion in the comment actions design doc was created?

Well the design doc was pretty light on the technical details for [comment linking](https://docs.google.com/document/d/1vfTkJEOqCu0bVKP9ATKhMqAfLslYMuoL6W80lLCB5U4/edit#bookmark=id.8sti1628sb6m). So from just that high-level idea, it hasn't changed much. But comment linking has a lot of nuances that I feel should be discussed either in more detail in a doc or in a GH plan of action comment.

Sure okay, I can ask CMEs. I'm not sure why to post in #expensify-open-source just yet though since atm the task is internal. If CMEs decide its external then I think it makes more sense to post in #expensify-open-source.

**Comment by mallenexpensify** [comment]
I think I _maybe_ accidentally said #expensify-open-source when I meant #engineering. 

If there hasn't been a detailed section written to specifically address comment linking, it sounds like it's be useful.  Maybe it'll help with splitting up internal/external issues too

**Comment by mallenexpensify** [comment]
My understanding is this issue is still open since we can't _prove_ everything works with bidirectional scrolling yet, even through it should. [ This is the main tracking issue for Comment Linking](https://github.com/Expensify/Expensify/issues/147480), we _could_ close this then move discussion there, if/when needed.   ¯\_(ツ)_/¯

**Comment by JmillsExpensify** [comment]
@roryabraham @chiragsalian I stumbled upon this issue again and I still feel we should close it since nothing is actually happening here. As noted, we already have an issue for comment linking in https://github.com/Expensify/Expensify/issues/147480. Re-open if you disagree.
